### PR TITLE
Route finding

### PR DESCRIPTION
**Pull request by hc09141**
Provides priority queue and Dijkstra's algorithm for determining the best series of routes for a piece of mail using lowest cost as the heuristic. The algorithm also checks the mail is below the maximum weight and volume thresholds for a company route before considering it

### CODE REVIEW

File: app/models/company_route.rb [hunk 1]
```diff
@@ -15,4 +15,18 @@ class CompanyRoute < ActiveRecord::Base
   validates :frequency, presence: true
   validates :duration, presence: true
   validates :day, presence: true, inclusion: {in: Constants::DAYS}
-end
+
+  def next_delivery_time(current_date_time)
+    week_day_number = current_date_time.wday
+    day_range = (week_day_number..6).to_a + (0...week_day_number).to_a
```
Comment (TanaTanoi) [inline app/models/company_route.rb]: maybe make 6 and 0 constants that relate to the actual day associated with it, like 0 = Monday (or Sunday?) etc.

File: app/models/create_mail.rb [hunk 2]
```diff
@@ -11,4 +12,54 @@ def to_s
     Priority:     #{priority}
     -----------------------------"
   end
+
+  def find_best_route()
```
Comment (TanaTanoi) [inline app/models/create_mail.rb]: can remove brackets here cos no params

File: app/models/create_mail.rb [hunk 3]
```diff
@@ -11,4 +12,54 @@ def to_s
...
+
+      if current.node.location == destination
+        return generate_best_route(current)
+      end
+
+      if !current.node.visited
+        current.node.visited = true
+        neighbours = CompanyRoute.where(origin: current.node.location)
+        neighbours.each do |route|
+          # requires mail to below max weight and volume for route
```
Comment (TanaTanoi) [inline app/models/create_mail.rb]: noice, appropriate place for comment

File: app/models/create_mail.rb [hunk 4]
```diff
@@ -11,4 +12,54 @@ def to_s
...
+              add_route_to_fringe(fringe, neighbour_node, current, route)
+            end
+          end
+        end
+      end
+    end
+  end
+
+  def add_route_to_fringe(fringe, neighbour_node, current, route)
+    fringe.push(Edge.new(node: neighbour_node, from: current.node,
```
Comment (TanaTanoi) [inline app/models/create_mail.rb]: I'd put each keyword argument on a new line to make it look nice (tbh the two per line looks 'aight, but i'd still put the last bracket on a new line)

File: app/models/delivery_time.rb [hunk 5]
```diff
@@ -9,4 +9,8 @@ class DeliveryTime < ActiveRecord::Base
   # def ensure_company_route_exists
   #   errors.add('CompanyRoute') unless self.company_route.find_by_id(self.company_route_id)
   # end
+
+  def date_time
+    @date_time ||= DateTime.parse("#{day} #{time}")
```
Comment (TanaTanoi) [inline app/models/delivery_time.rb]: noice

**Comment by Recursively** [comment]
@hc09141 Great work on this. Were you planning on doing some basic integration tests or are we just scraping tests at this point?

**Comment by hc09141** [comment]
@Recursively I'm keen to write integration tests if someone is up to giving a quick intro to testing tomorrow or has a good reference for testing

**Comment by Recursively** [comment]
@hc09141 Sweet we can pair on some tests tomorrow if you'd like

**Pull request merged**